File: app/src/main/java/com/xpf/p2p/utils/WorkUtil.java
package com.xpf.p2p.utils;

import android.app.ActivityManager;
import android.content.Context;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by xpf on 2017/8/26 :)
 * Function:
 */

public class WorkUtil {

    private static final String TAG = WorkUtil.class.getSimpleName();

    /**
     * 检测service是否在运行
     *
     * @param context
     * @param serviceName
     * @return
     */
    public static boolean isServiceWorked(Context context, String serviceName) {
        ActivityManager myManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        ArrayList<ActivityManager.RunningServiceInfo> runningService =
                (ArrayList<ActivityManager.RunningServiceInfo>) myManager.getRunningServices(Integer.MAX_VALUE);
        Log.e(TAG, "当前运行的: " + runningService);
        for (int i = 0; i < runningService.size(); i++) {
            if (runningService.get(i).service.getClassName().equals(serviceName)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 判断应用是否在运行
     *
     * @param context
     * @return
     */
    public static boolean isRun(Context context) {
        if (context == null) return false;
        Context appContext = context.getApplicationContext();
        ActivityManager am = (ActivityManager) appContext.getSystemService(Context.ACTIVITY_SERVICE);
        List<ActivityManager.RunningTaskInfo> list = am.getRunningTasks(100);
        boolean isAppRunning = false;
        String MY_PKG_NAME = context.getPackageName();
        //100表示取的最大的任务数，info.topActivity表示当前正在运行的Activity，info.baseActivity表系统后台有此进程在运行
        for (ActivityManager.RunningTaskInfo info : list) {
            if (info.topActivity.getPackageName().equals(MY_PKG_NAME) || info.baseActivity.getPackageName().equals(MY_PKG_NAME)) {
                isAppRunning = true;
                Log.i(TAG, info.topActivity.getPackageName() + " info.baseActivity.getPackageName()=" + info.baseActivity.getPackageName());
                break;
            }
        }

        Log.i(TAG, "程序 isAppRunning......" + isAppRunning);
        return isAppRunning;
    }
}

File: app/src/main/java/com/xpf/p2p/utils/WorkUtil.kt
package com.xpf.p2p.utils

import android.content.Context
import android.util.Log

/**
 * Created by xpf on 2017/8/26 :)
 * Function:
 */
object WorkUtil {

    private val TAG = WorkUtil::class.java.simpleName

    @Suppress("DEPRECATION")
    @JvmStatic
    fun isServiceWorked(context: Context, serviceName: String): Boolean {
        val myManager = context.getSystemService(Context.ACTIVITY_SERVICE) as android.app.ActivityManager
        val runningService = myManager.getRunningServices(Int.MAX_VALUE) as ArrayList<android.app.ActivityManager.RunningServiceInfo>
        Log.e(TAG, "当前运行的: $runningService")
        for (i in runningService.indices) {
            if (runningService[i].service.className == serviceName) {
                return true
            }
        }
        return false
    }

    @Suppress("DEPRECATION")
    @JvmStatic
    fun isRun(context: Context?): Boolean {
        if (context == null) return false
        val appContext = context.applicationContext
        val am = appContext.getSystemService(Context.ACTIVITY_SERVICE) as android.app.ActivityManager
        val list = am.getRunningTasks(100)
        var isAppRunning = false
        val MY_PKG_NAME = context.packageName
        for (info in list) {
            if (info.topActivity!!.packageName == MY_PKG_NAME || info.baseActivity!!.packageName == MY_PKG_NAME) {
                isAppRunning = true
                Log.i(TAG, "${info.topActivity!!.packageName} info.baseActivity.getPackageName()=${info.baseActivity!!.packageName}")
                break
            }
        }
        Log.i(TAG, "程序 isAppRunning......$isAppRunning")
        return isAppRunning
    }
}
